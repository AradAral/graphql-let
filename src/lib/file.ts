import { promises as fsPromises } from 'fs';
import glob from 'globby';
import _rimraf from 'rimraf';
import { promisify } from 'util';

export const rimraf = promisify(_rimraf);

export const { readFile, writeFile } = fsPromises;

export async function removeOldCache(
  cwd: string,
  tsxRelRegex: string,
  dtsRelRegex: string,
) {
  // Erasing old cache in __generated__ on HMR.
  // Otherwise the multiple `declare module "*/x.graphql"` are exposed.
  const oldFiles = await glob([tsxRelRegex, dtsRelRegex], {
    cwd,
    absolute: true,
  });
  await Promise.all(oldFiles.map(f => rimraf(f)));
}
