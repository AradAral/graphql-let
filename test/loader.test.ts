/* eslint-disable @typescript-eslint/no-non-null-assertion */

import { deepStrictEqual, ok, strictEqual } from 'assert';
import glob from 'globby';
import { join as pathJoin } from 'path';

import compiler from './__tools/compile';
import { rimraf } from './__tools/file';

const fixturePath1 = pathJoin(__dirname, '__fixtures/loader/usual');
const fixturePath2 = pathJoin(__dirname, '__fixtures/loader/monorepo');

describe('graphql-let/loader', () => {
  beforeEach(async () => {
    await rimraf(pathJoin(fixturePath1, '__generated__'));
  });

  test('generates .tsx and .d.ts', async () => {
    const fixture = 'pages/viewer.graphql';
    const stats = await compiler(fixturePath1, fixture, 'node');
    const { 0: actual, length } = stats
      .toJson()
      .modules!.map((m) => m.source)
      .filter(Boolean);

    deepStrictEqual(length, 1);
    expect(actual).toMatchSnapshot();
  });

  test('runs well for simultaneous execution, assuming SSR', async () => {
    const expectedTargets: [string, 'node' | 'web'][] = [
      ['pages/viewer.graphql', 'node'],
      ['pages/viewer.graphql', 'web'],
      ['pages/viewer2.graphql', 'node'],
      ['pages/viewer2.graphql', 'web'],
    ];
    const results = await Promise.all(
      expectedTargets.map(([file, target]) =>
        compiler(fixturePath1, file, target),
      ),
    );
    for (const [i, stats] of results.entries()) {
      const [file] = expectedTargets[i];
      const { 0: actual, length } = stats
        .toJson()
        .modules!.map((m) => m.source)
        .filter(Boolean);

      deepStrictEqual(length, 1);
      switch (file) {
        case 'pages/viewer.graphql':
          ok(actual!.includes('export function useViewerQuery('));
          break;
        case 'pages/viewer2.graphql':
          ok(actual!.includes('export function useViewer2Query('));
          break;
      }
    }
    const globResults = await glob('**/*.graphql.d.ts', { cwd: fixturePath1 });
    strictEqual(globResults.length, 2);
  });

  test('accepts config path in `options.configFile`', async () => {
    const stats = await compiler(
      pathJoin(fixturePath2, 'packages/app'),
      'src/index.ts',
      'web',
      { configFile: '../../config/.graphql-let.yml' },
    );

    const generated = stats
      .toJson()
      .modules?.flatMap((m) => m.modules)
      .find((m) => m?.name === './src/fruits.graphql');

    expect(generated?.source).toContain('export function useGetFruitsQuery');
    expect(
      generated?.source?.replace(/\/\*[\s\S]*?\*\//g, ''),
    ).toMatchSnapshot();
  });
});
