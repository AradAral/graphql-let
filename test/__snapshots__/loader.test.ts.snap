// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql-let/loader generates files by .graphql 1`] = `
Array [
  "import * as Apollo from '@apollo/client';
const gql = Apollo.gql;
/** All built-in and custom scalars, mapped to their actual values */

export const ViewerDocument = gql\`
    query Viewer {
  viewer {
    id
    name
    status
  }
}
    \`;
/**
 * __useViewerQuery__
 *
 * To run a query within a React component, call \`useViewerQuery\` and pass it any options that fit your needs.
 * When your component renders, \`useViewerQuery\` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useViewerQuery({
 *   variables: {
 *   },
 * });
 */

export function useViewerQuery(baseOptions) {
  return Apollo.useQuery(ViewerDocument, baseOptions);
}
export function useViewerLazyQuery(baseOptions) {
  return Apollo.useLazyQuery(ViewerDocument, baseOptions);
}",
]
`;

exports[`graphql-let/loader generates files by .tsx 1`] = `
Array [
  "import * as Vf9caa18689fd5fc145b366e28630414771b7f5a0 from \\"./../__generated__/pages/index-f9caa18689fd5fc145b366e28630414771b7f5a0.tsx\\";
const {
  useViewerQuery
} = Vf9caa18689fd5fc145b366e28630414771b7f5a0;
export default function Viewer() {
  const {
    data
  } = useViewerQuery();
  if (data) return /*#__PURE__*/React.createElement(\\"div\\", null, data.viewer.name);
}",
  "/* f9caa18689fd5fc145b366e28630414771b7f5a0
 * This file is automatically generated by graphql-let. */
import * as Apollo from '@apollo/client';
const gql = Apollo.gql;
/** All built-in and custom scalars, mapped to their actual values */

export const ViewerDocument = gql\`
    query Viewer {
  viewer {
    name
  }
}
    \`;
/**
 * __useViewerQuery__
 *
 * To run a query within a React component, call \`useViewerQuery\` and pass it any options that fit your needs.
 * When your component renders, \`useViewerQuery\` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useViewerQuery({
 *   variables: {
 *   },
 * });
 */

export function useViewerQuery(baseOptions) {
  return Apollo.useQuery(ViewerDocument, baseOptions);
}
export function useViewerLazyQuery(baseOptions) {
  return Apollo.useLazyQuery(ViewerDocument, baseOptions);
}",
]
`;
